source/tutorials/hpc.rst

```diff
@@ -0,0 +1,693 @@
+   
+   Example output:
+     
+   .. code-block:: console
+
+      hpc-worker4
+      hpc-worker3
+      hpc-worker1
+      hpc-worker2
+
+Create and run example scripts
+******************************
+
+Example 1: Return the hostname of each worker and output to :file:`show-hostnames.out`
+======================================================================================
+
+#. On the controller node, create the slurm :file:`show-hostnames.sh` script.
+
+   cat > show-hostnames.sh << EOF
+   #!/bin/bash
+   #
+   #SBATCH --job-name=show-hostnames
+   #SBATCH --output=show-hostnames.out
+   #
+   #SBATCH --ntasks=4
+   #SBATCH --time=10:00
+   #SBATCH --mem-per-cpu=100
+   #SBATCH --ntasks-per-node=1
+
+   srun hostname
+   EOF
+
+#. Execute the script.
+
+   .. code-block:: bash
+
+      sbatch show-hostnames.sh
+   
+   The result will be appear on first node.  
+
+#. View the result.
+
+   .. code-block:: bash
+
+      pdsh -w hpc-worker1 "cat show-hostnames.out"
+
+   Example output:
+
+   .. code-block:: console
+
+      hpc-worker1: hpc-worker3
+      hpc-worker1: hpc-worker4
+      hpc-worker1: hpc-worker1
+      hpc-worker1: hpc-worker2
+
+Example 2: An MPI "Hello, World!" program
+=========================================
+
+#. On the controller node, create the :file:`mpi-helloworld.c` program.
+
+   .. code-block:: bash
+
+      cat > mpi-helloworld.c << EOF 
+      #include <stdio.h>
+      #include <unistd.h>
+      #include <mpi.h>
+
+      int main(int argc, char** argv)
+      {
+        // Init the MPI environment
+        MPI_Init(NULL, NULL);
+
+        // Get the number of processes
+        int world_size;
+        MPI_Comm_size(MPI_COMM_WORLD, &world_size);
+
+        // Get the rank of the process
+        int world_rank;
+        MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
+
+        // Get the name of the processor
+        char processor_name[MPI_MAX_PROCESSOR_NAME];
+        int name_len;
+        MPI_Get_processor_name(processor_name, &name_len);
+
+        // Print a hello world message
+        printf("Hello, World! from from processor %s, rank %d out of %d processors\n", processor_name, world_rank, world_size);
+
+        // Finalize the MPI environment
+        MPI_Finalize();
+     }
+     EOF
+
+#. Add the `c-basic` and `devpkg-openmpi` bundles, which are needed to compile
+   it.
+
+   .. code-block:: bash
+
+      sudo swupd bundle-add c-basic devpkg-openmpi
+
+#. Compile it.
+   
+   .. code-block:: bash
+
+      mpicc -o mpi-helloworld mpi-helloworld.c
+
+#. Copy the binary to all worker nodes.
+
+   .. code-block:: bash
+
+      pdcp -P workers ./mpi-helloworld $HOME
+
+#. Create a slurm batch script to run it.
+
+   .. code-block:: bash
+
+      cat > mpi-helloworld.sh << EOF
+      #!/bin/sh
+      #SBATCH -o mpi-helloworld.out
+      #SBATCH --nodes=4
+      #SBATCH --ntasks-per-node=1
+
+      srun ./mpi-helloworld
+      EOF
+
+#. View the results on first worker node.
+
```

Comment:
As above:  the first worker node in the partition.

Reply:
@jurobystricky - Out of curiosity, does the order the workers in a partition matter?

For example: 
PartitionName=workers Nodes=hpc-worker1, hpc-worker2, hpc-worker3 Default=YES MaxTime=INFINITE State=UP

vs. 

PartitionName=workers Nodes=hpc-worker2, hpc-worker1, hpc-worker3 Default=YES MaxTime=INFINITE State=UP

In the first instance, the first worker is hpc-worker1.
In the 2nd instance, the first worker is hpc-worker3.